### PR TITLE
Beam-2383 Round function

### PR DESCRIPTION
**Pull request by app-tarush**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by app-tarush** [comment]
@XuMingmin please review this PR #3326

**Comment by aaltay** [comment]
R: @XuMingmin @takidau

**Comment by XuMingmin** [comment]
@xumingming can you take a peak here?

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlRoundExpression.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        break;
+      case DECIMAL:
+        result = BeamSqlPrimitive.of(SqlTypeName.DECIMAL,
+            roundBigDecimal(toBigDecimal(leftOp.getValue()), toInt(rightOp.getValue())));
+        break;
+    }
+    return result;
+  }
+
+  private int roundInt(int v1, int v2) {
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlRoundExpression.java]: I'm not sure I understand what purpose these round* methods serve. They obscure the fact that SqlFunctions.sround is the same (overloaded) method being used in all cases, but don't really seem to provide much. Do they just exist to have shorter names?

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    List<BeamSqlExpression> operands = new ArrayList<>();
+
+    // varchar operand not allowed
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.VARCHAR, "2"));
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.INTEGER, 4));
+    Assert.assertFalse(new BeamSqlRoundExpression(operands).accept());
+  }
+
+  @Test public void testRoundFunction() {
+    // operands are of type long, int, tinyint
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: This comment doesn't make sense to me. The operands immediately following are doubles.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    List<BeamSqlExpression> operands = new ArrayList<>();
+
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.DOUBLE, 2.0));
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.DOUBLE, 4.0));
+    assertEquals(2.0, new BeamSqlRoundExpression(operands).evaluate(record).getValue());
+    // operands are of type decimal, double, big decimal
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: Ditto.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    assertEquals(2,
+        new BeamSqlRoundExpression(operands).evaluate(record).getValue());
+
+    // operand with a BeamSqlInputRefExpression
+    operands.clear();
+    BeamSqlInputRefExpression ref0 = new BeamSqlInputRefExpression(SqlTypeName.BIGINT, 0);
+    operands.add(ref0);
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.BIGINT, 2L));
+
+    assertEquals(1234567L, new BeamSqlRoundExpression(operands).evaluate(record).getValue());
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: I'm confused by this one. Where is 1234567 coming from?

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlRoundExpression.java (hunk 1 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlRoundExpression.java]: @takidau they are to serve shorter names you got it correct

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java (hunk 2 above)
Comment (app-tarush) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: @takidau I will correct it to make it more meaningfull

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java (hunk 3 above)
Comment (app-tarush) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: @takidau I will correct it to make it more meaningfull

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java (hunk 4 above)
Comment (app-tarush) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: @takidau this refers to BeamSqlInputRefExpression which means first index value we are referencing. Lets suppose I have one row which represents a record with 4 columns

```java
record.addField(0, 1234567L);
    record.addField(1, 0);
    record.addField(2, 8.9);
    record.addField(3, 1234567L);
``` 

now BeamSqlInputRefExpression helps in referencing a column : 

```java
BeamSqlInputRefExpression ref0 = new BeamSqlInputRefExpression(SqlTypeName.BIGINT, 0);
```

```java
public class BeamSQLFnExecutorTestBase {
  public static RexBuilder rexBuilder = new RexBuilder(BeamQueryPlanner.TYPE_FACTORY);
  public static RelOptCluster cluster = RelOptCluster.create(new VolcanoPlanner(), rexBuilder);

  public static final JavaTypeFactory TYPE_FACTORY = new JavaTypeFactoryImpl(
      RelDataTypeSystem.DEFAULT);
  public static RelDataType relDataType;

  public static BeamSQLRecordType beamRecordType;
  public static BeamSQLRow record;

  public static RelBuilder relBuilder;

  @BeforeClass
  public static void prepare() {
    relDataType = TYPE_FACTORY.builder()
        .add("order_id", SqlTypeName.BIGINT)
        .add("site_id", SqlTypeName.INTEGER)
        .add("price", SqlTypeName.DOUBLE)
        .add("order_time", SqlTypeName.BIGINT).build();

    beamRecordType = BeamSQLRecordType.from(relDataType);
    record = new BeamSQLRow(beamRecordType);

    record.addField(0, 1234567L);
    record.addField(1, 0);
    record.addField(2, 8.9);
    record.addField(3, 1234567L);

    SchemaPlus schema = Frameworks.createRootSchema(true);
    final List<RelTraitDef> traitDefs = new ArrayList<RelTraitDef>();
    traitDefs.add(ConventionTraitDef.INSTANCE);
    traitDefs.add(RelCollationTraitDef.INSTANCE);
    FrameworkConfig config = Frameworks.newConfigBuilder()
        .parserConfig(SqlParser.configBuilder().setLex(Lex.MYSQL).build()).defaultSchema(schema)
        .traitDefs(traitDefs).context(Contexts.EMPTY_CONTEXT).ruleSets(BeamRuleSets.getRuleSets())
        .costFactory(null).typeSystem(BeamRelDataTypeSystem.BEAM_REL_DATATYPE_SYSTEM).build();

    relBuilder = RelBuilder.create(config);
  }
}
```

Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: Ah, got it, thanks!

**Comment by app-tarush** [comment]
@takidau I have made changes in comments.

**Comment by takidau** [comment]
To clarify, are you wanting to pull this change before https://github.com/apache/beam/pull/3336, or are you wanting to submit POWER and ROUND together with just that new PR? I think this one is probably ready to go from my perspective.

**Comment by app-tarush** [comment]
@takidau I am good if this pull with change #3326 can go before POWER #3336.

**Comment by takidau** [comment]
Great, I'll go ahead and pull this one then. LGTM.

**Comment by app-tarush** [comment]
@takidau thanks!! 👍

**Comment by takidau** [comment]
Merged on branch DSL_SQL. I updated to include the SQL → Sql renames as part of it. Feel free to close this PR. Thanks!

**Comment by app-tarush** [comment]
@takidau thanks!!

**Pull request closed without merging**